src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -3042,56 +3042,85 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         _core.ClearHoveredCell();
     }
 
-    winrt::fire_and_forget TermControl::_hoveredHyperlinkChanged(IInspectable /*sender*/,
-                                                                 IInspectable /*args*/)
+    // Attackers abuse Unicode characters that happen to look similar to ASCII characters. Cyrillic for instance has
+    // its own glyphs for а, с, е, о, р, х, and у that look practically identical to their ASCII counterparts.
+    // This is called a "IDN homoglyph attack".
+    //
+    // But outright showing Punycode URIs only is similarly flawed as they can end up looking similar to valid ASCII URIs.
+    // xn--cnn.com for instance looks confusingly similar to cnn.com, but actually represents U+407E.
+    //
+    // An optimal solution would detect any URI that contains homoglyphs and show them in their Punycode form.
+    // Such a detector however is not quite trivial and requires constant maintenance, which this project's
+    // maintainers aren't currently well equipped to handle. As such we do the next best thing and show the
+    // Punycode encoding side-by-side with the Unicode string for any IDN.
+    static winrt::hstring sanitizeURI(winrt::hstring uri)
     {
-        auto weakThis{ get_weak() };
-        co_await wil::resume_foreground(Dispatcher());

```

Comment:
This is another instance where we were switching to the foreground to make touching UI objects safe. _Are we convinced that this is safe?_

Reply:
The only place `_hoveredHyperlinkChanged` is called from is the `_core.HoveredHyperlinkChanged()` event. `_HoveredHyperlinkChangedHandlers` is only called in `ControlCore::_updateHoveredCell`, which in turn is only called by `SetHoveredCell` and `ClearHoveredCell`. The former is only called by `ControlInteractivity::PointerMoved` which is a UI-thread-only function and the latter is only called by `TermControl::_PointerExitedHandler` which is the same. BTW It's somewhat weird that `PointerMoved` us in `ControlInteractivity`, but "pointer exited" isn't.

So yeah this can only be called from the UI thread.

Reply:
That's technically an implementation detail, but I'll allow it ;P